### PR TITLE
Fix queue time calculation for builds and pipeline steps

### PR DESCRIPTION
**Pull request by nikita-tkachenko-datadog**
### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/CONTRIBUTING.md). 

### What does this PR do?

Fixes the logic that calculates queue time for pipelines and pipeline steps.

The following problems are fixed:
- queue time calculated incorrectly for pipelines that first wait to start (for example, because of throttling settings) and then wait for their executor to become available. These waiting times should be summed (because the two waits happen sequentially), while the old implementations only used the biggest of the two values.
- pipeline span and all of its children spans are shifted into the past, because propagated queue time is subtracted from both the start and the end timestamps (I think it used to be done this way because the plugin used DD Java Agent's API for creating spans. Pipeline span had to start before queue time, and modifying start time later was not possible)
- queueing time of individual pipeline steps is counted twice: first as the queuing time, and second as part of the step itself (both durations can be seen in the pipeline trace)

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java [hunk 1]
```diff
@@ -46,15 +45,13 @@ public JSONObject toJson(PipelineStepData flowNode, Run<?, ?> run) throws IOExce
...
-        final long propagatedMillisInQueue = Math.max(buildData.getPropagatedMillisInQueue(-1L), 0);
-        final long fixedStartTimeNanos = TimeUnit.MICROSECONDS.toNanos(current.getStartTimeMicros() - TimeUnit.MILLISECONDS.toMicros(propagatedMillisInQueue));
-        final long fixedEndTimeNanos = TimeUnit.MICROSECONDS.toNanos(current.getEndTimeMicros() - TimeUnit.MILLISECONDS.toMicros(propagatedMillisInQueue));
+        final long startTimeNanos = TimeUnit.MILLISECONDS.toNanos(current.getStartTimeMillis());
+        final long endTimeNanos = TimeUnit.MILLISECONDS.toNanos(current.getEndTimeMillis());
 
         // At this point, the current node is traceable.
         final TraceSpan.TraceSpanContext spanContext = new TraceSpan.TraceSpanContext(current.getTraceId(), current.getParentSpanId(), current.getSpanId());
-        final TraceSpan span = new TraceSpan(buildOperationName(current), fixedStartTimeNanos + current.getNanosInQueue(), spanContext);
+        final TraceSpan span = new TraceSpan(buildOperationName(current), startTimeNanos + current.getQueueTimeMillis(), spanContext);
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java:53]: Are we removing the queue time twice? We're doing the same [in this part of the code](https://github.com/jenkinsci/datadog-plugin/blob/d2537c85e514e669736c53aed5a75c587d5740e5/src/main/java/org/datadog/jenkins/plugins/datadog/model/PipelineStepData.java#L173) and [we're not doing it for webhooks protocol.](https://github.com/jenkinsci/datadog-plugin/blob/d2537c85e514e669736c53aed5a75c587d5740e5/src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogWebhookPipelineLogic.java#L55)

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildData.java [hunk 2]
```diff
@@ -135,8 +135,32 @@ public class BuildData implements Serializable {
...
+     * Queue time that is propagated from a child node of a pipeline:
+     * if a pipeline is configured to execute in a specific executor (with the top-level `agent {}` section),
+     * it will be reported as started even if that executor is not available.
+     * In this case the time spent waiting for the executor will be reported by the first child node of the pipeline,
+     * and propagated to the pipeline from there.
+     * <p>
+     * This queue time goes AFTER timestamp that is reported by {@link Run#getStartTimeInMillis()}
+     * and is INCLUDED into the duration reported by {@link Run#getDuration()}.
+     */
+    private long propagatedMillisInQueue;
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildData.java:163]: Do we know how does this work when the pipeline uses multiple agents? 
https://www.jenkins.io/doc/book/pipeline/jenkinsfile/#using-multiple-agents

File: src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogBuildListener.java [hunk 3]
```diff
@@ -218,26 +237,14 @@ public void onStarted(Run run, TaskListener listener) {
...
-            try (Metrics metrics = client.metrics()) {
-                long waitingMs = (DatadogUtilities.currentTimeMillis() - item.getInQueueSince());
-                metrics.gauge("jenkins.job.waiting", TimeUnit.MILLISECONDS.toSeconds(waitingMs), hostname, tags);
 
-                PipelineQueueInfoAction queueInfoAction = run.getAction(PipelineQueueInfoAction.class);
-                if (queueInfoAction != null) {
-                    queueInfoAction.setQueueTimeMillis(waitingMs);
+            if (waitingMs != null) {
+                try (Metrics metrics = client.metrics()) {
+                    metrics.gauge("jenkins.job.waiting", TimeUnit.MILLISECONDS.toSeconds(waitingMs), hostname, tags);
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogBuildListener.java:246]: Not sure if this metric has the same meaning than CI Visibility Queue Time, but here we're not taking into account the "propagated" queue time (for Jenkins Declarative Pipelines using `agent`)

**Review by drodriguezhdez — COMMENTED** [review]
Dropped some comments.

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildData.java (hunk 2 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildData.java:163]: As `BuildData` represent the entire build, only time spent waiting for the agent specified in the top-level `agent {}` section is propagated, which I think is the correct behaviour. 
For individual stages the time spent waiting for their respective agents will be accounted for in the `PipelineStepData` instances

File: src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogBuildListener.java (hunk 3 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogBuildListener.java:246]: I'm not sure either, the documentation just says "Time spent waiting for job to run".
I'd leave it as is, at least for now, given that there haven't been any customer complaints

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java (hunk 1 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java:53]: Good catch, fixed!

**Review by drodriguezhdez — APPROVED** [review]
LGTM

**Pull request merged**